### PR TITLE
Add Loki-operator to STO

### PR DESCRIPTION
**Pull request by vyzigold**
This PR adds possibility to deploy STO with loki-operator and deploy Loki and log SG with it.

A few basic configuration options for the loki-operator are exposed through STO configuration. More options can be easily added later when needed.

Build and deployment is temporarily added to the stf-run-ci ansible role for easier development. This won't be needed once the operator is released. The loki-operator is currently under development, so I made a few workarounds to make some things work.
The main workarounds are:

- Remove authentication from /metrics - this seemed to prevent loki microservices from communicating with each other
- Remove operator-sdk build - I wasn't able to build operator-sdk, so I'm downloading a prebuilt binary instead.
- Skip tls verification - There doesn't seem to be an option to verify OCS certificate yet, so the config file is patched to skip it.

I added examples for deploying with minio to crc and with OCS to something like PSI to the docs.

### CODE REVIEW

File: build/stf-run-ci/tasks/main.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/main.yml:61]:
```suggestion
    - name: Skip Loki TLS verification
```

File: deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml:164]: Generally `size` is used to indicate the number of replicas. I think this might be better as `flavor` or something like that.

File: deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml:222]:
```suggestion
```

This is a duplicate stanza

**Review by leifmadsen — COMMENTED** [review]
This looks really really good for a first pass. Nice work! Couple minor comments. I'd like to go over this one more time before approving it, but nothing jumps out at me that should prevent this from merging shortly. Thanks for updating the documentation too! I'd like to see a demo of this working, but so far so good!

**Comment by vyzigold** [comment]
Thank you for the review, I commited your suggestions. I'm not here next week, but I guess I could prepare some demo of this for the happy hour on August 25th.

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (csibbitt) [inline build/stf-run-ci/tasks/clone_repos.yml:68]: Note that this branch logic isn't going to work quite like the other components because in general if we were to make a  PR against the loki-operator, it will not be a branch on the viaq repo since we don't have write access to it. It still might be useful, and it's not dangerous, but I wanted to point this out.

File: roles/servicetelemetry/defaults/main.yml
Comment (csibbitt) [inline roles/servicetelemetry/defaults/main.yml:50]: I'm not finding this used anywhere - is this supposed to be `replication_factor`?

**Review by csibbitt — APPROVED** [review]
Overall LGTM but I haven't had a chance to test it yet.

File: build/stf-run-ci/tasks/main.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/main.yml:51]:
```suggestion
    # TLS verification support doesn't seem to be implemented in the operator yet
```

File: build/stf-run-ci/tasks/main.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/main.yml:61]:
```suggestion
    - name: Skip Loki TLS verification
```

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (leifmadsen) [reply build/stf-run-ci/tasks/clone_repos.yml:68]: Might be worth just adding a little comment in the file about that. I suspect it's generally fine just because I'm not sure how much we'll be doing PRs and needing to do tests against it, but obviously that won't work with this repo.

**Review by leifmadsen — APPROVED** [review]

File: roles/servicetelemetry/defaults/main.yml
Comment (vyzigold) [reply roles/servicetelemetry/defaults/main.yml:50]: yes

Comment (vyzigold) [reply roles/servicetelemetry/defaults/main.yml]: I changed it to replication_factor

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (vyzigold) [reply build/stf-run-ci/tasks/clone_repos.yml:71]: I added a note about this to the file

**Review by pleimer — APPROVED** [review]

File: build/get_new_operator_sdk.sh
Comment (leifmadsen) [inline build/get_new_operator_sdk.sh:6]: I feel like this would be better if you made the version a variable that could be overridden and then used instead of hard coding it across the file.

File: build/stf-run-ci/tasks/main.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/main.yml:44]: If you make the version configurable in the script, I'd pass an argument here with the version you expect.

File: build/stf-run-ci/tasks/setup_stf_local_build.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/setup_stf_local_build.yml:93]: If you set the version as a fact in the defaults you could use it here and in the other plays to sync the version across. I'm not sure I like seeing the version hard coded across so many locations.

File: build/stf-run-ci/tasks/setup_stf_local_build.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/setup_stf_local_build.yml:149]: Hard-coded reference

File: build/get_new_operator_sdk.sh
Comment (vyzigold) [reply build/get_new_operator_sdk.sh:6]: Yes, this sounds better

**Comment by vyzigold** [comment]
@leifmadsen I made the operator-sdk version a variable. I also noticed a few minor issues, which I fixed.

File: build/get_new_operator_sdk.sh
Comment (leifmadsen) [inline build/get_new_operator_sdk.sh:6]: You can set a default if you want here so that you can use the script as you were previously. This keeps it at 1.5.0 as your default, but at least it can be reused this way too.

```suggestion
VERSION="${1:-1.5.0}"
```

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/clone_repos.yml:69]: I'm not suggesting you change this now, but we could possibly make this a default parameter that could be overridden, and then if we forked it for the purposes of development, then at least you could consume that forked version while doing the development.

This is basically the case for all the repositories though, not just this one (although of course we as developers do have access to custom branches on the repo). Something we can deal with if this becomes a use case in the future though.

File: build/stf-run-ci/tasks/deploy_stf.yml
Comment (leifmadsen) [inline build/stf-run-ci/tasks/deploy_stf.yml:37]: I think this is supposed to be `storageClassName`?

File: deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml:177]: I noted earlier that you're calling `storageClass` in the manifest, but that you've defined this as `storageClassName` here. I think actually this should be renamed to `storageClass` liek the other `storage...` properties in this file for consistency.

File: build/stf-run-ci/tasks/deploy_stf.yml
Comment (leifmadsen) [reply build/stf-run-ci/tasks/deploy_stf.yml:37]: Further review, I think this is going to be correct and that the `storage.storageClassName` parameter should change to be consistent with other `storage...` parameters.

File: deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:66]: This will need to change as well for consistency.

File: deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:404]: Will need to change for consistency

File: docs/loki.md
Comment (leifmadsen) [inline docs/loki.md:24]: This one is fine because it's a separate object (not the ServiceTelemetry object)

File: roles/servicetelemetry/defaults/main.yml
Comment (leifmadsen) [inline roles/servicetelemetry/defaults/main.yml:54]: This one looks fine once made consistent across the `ServiceTelemetry` manifest.

**Review by leifmadsen — CHANGES_REQUESTED** [review]

**Comment by leifmadsen** [comment]
Sorry that I missed the consistency stuff the first time through, but this looks really good. I think once those consistency things are fixed up this is going to be good to go.

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (vyzigold) [reply build/stf-run-ci/tasks/clone_repos.yml:69]: I might have some time later this sprint, so I may take a look at this.

File: build/stf-run-ci/tasks/deploy_stf.yml
Comment (vyzigold) [reply build/stf-run-ci/tasks/deploy_stf.yml:37]: Interesting. I remember changing the storage class to standard2 here at some point and it seemed to work :thinking:. I'll fix it today.

File: build/stf-run-ci/tasks/clone_repos.yml
Comment (vyzigold) [reply build/stf-run-ci/tasks/clone_repos.yml:69]: The time came earlier than I thought :D

**Comment by vyzigold** [comment]
@leifmadsen I fixed the storageClass inconsistency.

File: deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:371]:
```suggestion
```

Comment (leifmadsen) [reply deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:371]: Just remove these extra newlines please

**Comment by leifmadsen** [comment]
Two really minor nit-pick things, then I think this is good to merge. Thanks!

**Comment by vyzigold** [comment]
Done @leifmadsen

**Review by leifmadsen — APPROVED** [review]

**Comment by leifmadsen** [comment]
test

File: deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml
Comment (leifmadsen) [inline deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:69]: I think this is causing an issue. I just noticed that there isn't a corresponding change to the `deploy/crds/infra.watch_v1beta1_servicetelemetry_cr.yaml` is missing here. Normally when you build the local bundle it should take the `CR` example file and will generate the example in the generated bundle CRD.

There needs to be a command run by `operator-sdk` that I'll look in to here. It's been a while since I've done it. May have a change incoming before we can merge this since CI is failing.

Comment (leifmadsen) [reply deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:69]: Yea confirmed. Changes that show up in the bundle also need to be put in to the `deploy/crds/` files, which are ultimately the source when you generate the bundle.

You'll want to run this command: `operator-sdk generate bundle --default-channel unstable --channels unstable` which will show you the changes being removed from the `deploy/olm-catalog/` because they aren't sourced in the `deploy/crds/` directory files.

Comment (leifmadsen) [reply deploy/olm-catalog/service-telemetry-operator/manifests/service-telemetry-operator.clusterserviceversion.yaml:69]: Here is a small patch for part of the changes because I was testing locally:

```diff
diff --git a/deploy/crds/infra.watch_v1beta1_servicetelemetry_cr.yaml b/deploy/crds/infra.watch_v1beta1_servicetelemetry_cr.yaml
index 7752eed..f84c060 100644
--- a/deploy/crds/infra.watch_v1beta1_servicetelemetry_cr.yaml
+++ b/deploy/crds/infra.watch_v1beta1_servicetelemetry_cr.yaml
@@ -34,6 +34,14 @@ spec:
           persistent:
             storageSelector: {}
             pvcStorageRequest: 20Gi
+    logs:
+      loki:
+        enabled: true,
+        flavor: 1.extra-small
+        replicationFactor: 1
+        storage:
+          objectStorageSecret: test
+          storageClass: standard
   graphing:
     enabled: false
     grafana:
```

File: deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml
Comment (csibbitt) [inline deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml:159]: Can we just change this description to say `---TESTING ONLY--- Enable Loki as a storage backend for logs` ?

Comment (leifmadsen) [reply deploy/olm-catalog/service-telemetry-operator/manifests/infra.watch_servicetelemetrys_crd.yaml:159]: Alternatively we could downstream patch this out during build, but that feels heavy?

**Comment by vyzigold** [comment]
CI doesn't work for PRs from forked repositories. Since this PR seems to be approved by everyone and the CI is is the only thing missing, I created a copy of this PR at #252. With this I'm closing this PR.

**Pull request closed without merging**
